File: examples/gethostlatency.c
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>
struct start_timestamp_t {
    u32 pid;
    char comm[TASK_COMM_LEN];
    char host[80];
    u64 ts;
};
struct latency_event_t {
    u32 pid;
    u64 delta;
    char comm[TASK_COMM_LEN];
    char host[80];
};
BPF_HASH(start, u32, struct start_timestamp_t);
BPF_PERF_OUTPUT(events);
int do_entry(struct pt_regs *ctx) {
    if (!PT_REGS_PARM1(ctx))
        return 0;
    struct start_timestamp_t val = {};
    u32 pid = bpf_get_current_pid_tgid();
    if (bpf_get_current_comm(&val.comm, sizeof(val.comm)) == 0) {
        bpf_probe_read_user(&val.host, sizeof(val.host),
                       (void *)PT_REGS_PARM1(ctx));
        val.pid = bpf_get_current_pid_tgid();
        val.ts = bpf_ktime_get_ns();
        start.update(&pid, &val);
    }
    return 0;
}
int do_return(struct pt_regs *ctx) {
    struct start_timestamp_t *start_timestamp;
    struct latency_event_t perf_event = {};
    u64 delta;
    u32 pid = bpf_get_current_pid_tgid();
    u64 finish_timestamp = bpf_ktime_get_ns();
    start_timestamp = start.lookup(&pid);
    if (start_timestamp == 0)
        return 0; // missed start
    bpf_probe_read_kernel(&perf_event.comm, sizeof(perf_event.comm), start_timestamp->comm);
    bpf_probe_read_kernel(&perf_event.host, sizeof(perf_event.host), (void *)start_timestamp->host);
    perf_event.pid = start_timestamp->pid;
    perf_event.delta = finish_timestamp - start_timestamp->ts;
    events.perf_submit(ctx, &perf_event, sizeof(perf_event));
    start.delete(&pid);
    return 0;
}

File: examples/gethostlatency.rs
use bcc::perf_event::PerfMapBuilder;
use bcc::BccError;
use bcc::{Uprobe, Uretprobe, BPF};
use clap::{App, Arg};

use core::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::{fmt, ptr};

/*
 * Basic Rust clone of `gethostlatency`, from the iovisor tools.
 * https://github.com/iovisor/bcc/blob/master/tools/gethostlatency.py
 *
 * Prints the latency for getaddrinfo/gethostbyname(3) calls, which are a common part of DNS lookups.
 */

/*
 * Define the struct the BPF code writes in Rust
 * This must match the struct in `gethostlatency.c` exactly.
 * The important thing to understand about the code in `gethostlatency.c` is that it creates structs of
 * type `latency_event_t` and pushes them into a buffer where our Rust code can read them.
 */
#[repr(C)]
struct latency_event_t {
    pid: u32,
    delta: u64,
    comm: [u8; 16], // TASK_COMM_LEN
    host: [u8; 80],
}

impl fmt::Display for latency_event_t {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let now = time::OffsetDateTime::now_utc();
        let date = now.date();
        let time = now.time();
        let dt_string = format!(
            "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}.{:09}Z",
            date.year(),
            date.month() as u8,
            date.day(),
            time.hour(),
            time.minute(),
            time.second(),
            time.nanosecond(),
        );
        write!(
            f,
            "{:<9} {:<6} {:<16} {:>10.2} {}",
            dt_string,
            self.pid,
            get_string(&self.comm),
            (self.delta as f32) / 1e6,
            get_string(&self.host)
        )
    }
}

fn do_main(runnable: Arc<AtomicBool>) -> Result<(), BccError> {
    let matches = App::new("gethostlatency")
        .about("Prints the latency for getaddrinfo/gethostbyname(2) calls")
        .arg(
            Arg::with_name("pid")
                .long("pid")
                .short("p")
                .help("Only trace calls from this PID")
                .takes_value(true),
        )
        .get_matches();

    let pid: Option<i32> = matches
        .value_of("pid")
        .map(|v| v.parse::<i32>().expect("Invalid argument for PID"));

    // attach entry and return probes for common host lookup functions
    let libc_so_path = "/lib/x86_64-linux-gnu/libc.so.6";
    let code = include_str!("gethostlatency.c");
    let mut module = BPF::new(code)?;
    for symbol in &["getaddrinfo", "gethostbyname", "gethostbyname2"] {
        Uprobe::new()
            .handler("do_entry")
            .binary(libc_so_path)
            .symbol(symbol)
            .pid(pid)
            .attach(&mut module)?;
        Uretprobe::new()
            .handler("do_return")
            .binary(libc_so_path)
            .symbol(symbol)
            .pid(pid)
            .attach(&mut module)?;
    }

    // the "events" table is where the "function was called" events get sent
    let table = module.table("events")?;
    // install a callback to print out latency events when they happen
    let mut perf_map = PerfMapBuilder::new(table, perf_data_callback).build()?;
    // print a header
    println!(
        "{:<9} {:<6} {:<16} {:>10} {}",
        "TIME", "PID", "COMM", "LATms", "HOST"
    );
    // this `.poll()` loop is what makes our callback get called
    while runnable.load(Ordering::SeqCst) {
        perf_map.poll(200);
    }
    Ok(())
}

fn perf_data_callback() -> Box<dyn FnMut(&[u8]) + Send> {
    Box::new(|x| {
        println!("{}", parse_struct(x));
    })
}

fn parse_struct(x: &[u8]) -> latency_event_t {
    unsafe { ptr::read_unaligned(x.as_ptr() as *const latency_event_t) }
}

fn get_string(x: &[u8]) -> String {
    match x.iter().position(|&r| r == 0) {
        Some(zero_pos) => String::from_utf8_lossy(&x[0..zero_pos]).to_string(),
        None => String::from_utf8_lossy(x).to_string(),
    }
}

fn main() {
    let runnable = Arc::new(AtomicBool::new(true));
    let r = runnable.clone();
    ctrlc::set_handler(move || {
        r.store(false, Ordering::SeqCst);
    })
    .expect("Failed to set handler for SIGINT / SIGTERM");

    match do_main(runnable) {
        Err(x) => {
            eprintln!("Error: {}", x);
            std::process::exit(1);
        }
        _ => {}
    }
}
